File: backend/app/schemas/resume.py
from typing import Optional
from datetime import datetime
from pydantic import BaseModel, ConfigDict, Field
import uuid


class ResumeBase(BaseModel):
    title: str = Field(..., max_length=255)
    content: str
    is_primary: Optional[bool] = False


class ResumeCreate(ResumeBase):
    application_id: Optional[uuid.UUID] = None


class ResumeUpdate(BaseModel):
    title: Optional[str] = None
    content: Optional[str] = None
    is_primary: Optional[bool] = None


class ResumeResponse(ResumeBase):
    id: uuid.UUID
    user_id: uuid.UUID
    application_id: Optional[uuid.UUID] = None
    pdf_url: Optional[str] = None
    versions_count: int = 0
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# --- Resume Versions ---


class ResumeVersionCreate(BaseModel):
    content: str
    template: Optional[str] = "minimal"
    job_id: Optional[uuid.UUID] = None
    parent_version_id: Optional[uuid.UUID] = None
    source_type: Optional[str] = "base"
    tailoring_notes: Optional[str] = None
    score: Optional[int] = None
    status: Optional[str] = "completed"


class ResumeVersionResponse(BaseModel):
    id: uuid.UUID
    resume_id: uuid.UUID
    version_number: int
    content: str
    template: Optional[str] = None
    job_id: Optional[uuid.UUID] = None
    parent_version_id: Optional[uuid.UUID] = None
    source_type: str = "base"
    job_title: Optional[str] = None
    job_company: Optional[str] = None
    tailoring_notes: Optional[str] = None
    score: Optional[int] = None
    status: str
    created_at: datetime
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class ResumeVersionListResponse(BaseModel):
    versions: list[ResumeVersionResponse]
    total: int


# --- Resume Diff ---


class DiffSummary(BaseModel):
    added: int = 0
    removed: int = 0
    changed: int = 0


class StringListDiff(BaseModel):
    added: list[str] = []
    removed: list[str] = []
    preserved: list[str] = []


class TextDiff(BaseModel):
    from_text: Optional[str] = None
    to_text: Optional[str] = None
    changed: bool = False


class ExperienceItemDiff(BaseModel):
    key: str
    status: str  # "added", "removed", "modified", "unchanged"
    company: str
    title: str
    period_from: Optional[str] = None
    period_to: Optional[str] = None
    location_from: Optional[str] = None
    location_to: Optional[str] = None
    bullets_added: list[str] = []
    bullets_removed: list[str] = []
    bullets_preserved: list[str] = []


class ProjectItemDiff(BaseModel):
    name: str
    status: str  # "added", "removed", "modified", "unchanged"
    description_from: Optional[str] = None
    description_to: Optional[str] = None
    url_from: Optional[str] = None
    url_to: Optional[str] = None


class EducationItemDiff(BaseModel):
    key: str
    status: str  # "added", "removed", "modified", "unchanged"
    institution: str
    degree: str
    field_of_study_from: Optional[str] = None
    field_of_study_to: Optional[str] = None
    period_from: Optional[str] = None
    period_to: Optional[str] = None


class ResumeDiffSections(BaseModel):
    summary: TextDiff
    skills: StringListDiff
    experience: list[ExperienceItemDiff] = []
    projects: list[ProjectItemDiff] = []
    education: list[EducationItemDiff] = []


class ResumeDiffResponse(BaseModel):
    from_version_id: uuid.UUID
    to_version_id: uuid.UUID
    from_version_number: int
    to_version_number: int
    has_changes: bool
    summary: DiffSummary
    sections: ResumeDiffSections


# --- Resume Generation ---


class ResumeGenerationRequest(BaseModel):
    job_id: uuid.UUID
    parent_version_id: Optional[uuid.UUID] = None
    template: str = Field(
        default="minimal",
        description="Template name: minimal, technical, modern, corporate, compact",
    )
    provider: Optional[str] = None
    model: Optional[str] = None


class ResumeGenerationResponse(BaseModel):
    version_id: uuid.UUID
    status: str
    message: str

File: frontend/src/types/resume.ts
// Resume types
export interface Resume {
  id: string;
  user_id: string;
  title: string;
  content: string;
  pdf_url: string | null;
  is_primary: boolean;
  application_id: string | null;
  created_at: string;
  updated_at: string;
}

export type ResumeVersionSource = "base" | "tailored" | "revert";

export interface ResumeVersion {
  id: string;
  resume_id: string;
  version_number: number;
  content: string;
  template: string | null;
  job_id: string | null;
  parent_version_id: string | null;
  source_type: ResumeVersionSource;
  job_title?: string | null;
  job_company?: string | null;
  tailoring_notes: string | null;
  score: number | null;
  status: "pending" | "processing" | "completed" | "failed";
  created_at: string;
  updated_at: string | null;
}

export interface ResumeGenerationRequest {
  job_id: string;
  parent_version_id?: string | null;
  template: string;
  provider?: string;
  model?: string;
}

export interface TailoredResumeData {
  name: string;
  email: string;
  phone: string | null;
  location: string | null;
  linkedin: string | null;
  github: string | null;
  website: string | null;
  summary: string;
  experience: {
    title: string;
    company: string;
    location: string | null;
    period: string;
    bullets: string[];
  }[];
  projects: {
    name: string;
    description: string;
    url: string | null;
  }[];
  skills: string[];
  education: {
    institution: string;
    degree: string;
    field_of_study: string | null;
    period: string;
  }[];
}

// --- Resume Diff Types ---

export interface DiffSummary {
  added: number;
  removed: number;
  changed: number;
}

export interface StringListDiff {
  added: string[];
  removed: string[];
  preserved: string[];
}

export interface TextDiff {
  from_text: string | null;
  to_text: string | null;
  changed: boolean;
}

export interface ExperienceItemDiff {
  key: string;
  status: "added" | "removed" | "modified" | "unchanged";
  company: string;
  title: string;
  period_from: string | null;
  period_to: string | null;
  location_from: string | null;
  location_to: string | null;
  bullets_added: string[];
  bullets_removed: string[];
  bullets_preserved: string[];
}

export interface ProjectItemDiff {
  name: string;
  status: "added" | "removed" | "modified" | "unchanged";
  description_from: string | null;
  description_to: string | null;
  url_from: string | null;
  url_to: string | null;
}

export interface EducationItemDiff {
  key: string;
  status: "added" | "removed" | "modified" | "unchanged";
  institution: string;
  degree: string;
  field_of_study_from: string | null;
  field_of_study_to: string | null;
  period_from: string | null;
  period_to: string | null;
}

export interface ResumeDiffSections {
  summary: TextDiff;
  skills: StringListDiff;
  experience: ExperienceItemDiff[];
  projects: ProjectItemDiff[];
  education: EducationItemDiff[];
}

export interface ResumeDiffResponse {
  from_version_id: string;
  to_version_id: string;
  from_version_number: number;
  to_version_number: number;
  has_changes: boolean;
  summary: DiffSummary;
  sections: ResumeDiffSections;
}
